src/actions/receive-profile.js

```diff
@@ -1015,73 +993,56 @@ export async function checkIfWebChannelUsableForSymbolication(): Promise<boolean
-          // instance is running on a different host than the one that's specified in the
-          // preference `devtools.performance.recording.ui-base-url`.
-          // That's unusual, because we get here only for the /from-browser entry point,
-          // which is usually accessed from the new tab that the browser opens when a
-          // profile is captured - and that tab gets opened on the profiler instance that's
-          // configured in the base URL.
-          // We completely stop trying to connect to the browser and display an error message.
-          // (We don't fall back to the frame script, because we'd most likely display
-          // the waiting animation indefinitely and never connect to a frame script.)
-          throw new Error(oneLine`
-            This profiler instance was unable to connect to the
-            WebChannel. This usually means that it’s running on a
-            different host from the one that is specified in the
-            preference devtools.performance.recording.ui-base-url. If
-            you would like to capture new profiles with this instance, you can go to about:config
-            and change the preference. Error: ${e.name}: ${e.message}
-          `);
-        }
+      // Attempt to establish a connection to the browser.
+      // Disable the userAgent check by supplying a fake userAgent that
+      // pretends we're Firefox. This will make us attempt to establish
+      // a connection to the WebChannel even if we're running in the test
+      // suite.
+      const connectionStatus = await createBrowserConnection('Firefox/123.0');
+      switch (connectionStatus.status) {
+        case 'ESTABLISHED':
+          // Good. This is the normal case.
+          break;
+        // The other cases are error cases.
+        case 'NOT_FIREFOX':
+          throw new Error('/from-browser only works in Firefox browsers');
+        case 'WAITING':
+          throw new Error('unexpected');
+        case 'DENIED':
+          throw connectionStatus.error;
+        case 'TIMED_OUT':
+          throw new Error(
+            'Timed out when waiting for reply to WebChannel message'
+          );
+        default:
+          throw assertExhaustiveCheck(connectionStatus.status);
       }
 
+      // Now we know that connectionStatus.status === 'ESTABLISHED'.
+      const browserConnection = connectionStatus.browserConnection;
+
       // XXX update state to show that we're connected to the browser
 
       dispatch(waitingForProfileFromBrowser());
 
-      let geckoProfiler;
-      let rawGeckoProfile;
-
-      if (shouldUseWebChannel) {
-        rawGeckoProfile = await getProfileViaWebChannel();
-      } else {
-        const timeoutId = setTimeout(() => {
-          dispatch(
-            temporaryError(
-              new TemporaryError(oneLine`
+      await browserConnection.establishConnectionViaFrameScriptIfNeeded(() => {
```

Comment:
I find that at the call site, the role of the function isn't obvious... What do you think about using an option object rather than a function parameter? `browserConnection.establishConnectionViaFrameScriptIfNeeded({ onThirtySecondTimeout: XXX })`

Reply:
I agree. I was considering doing that but didn't know if it was worth doing. Thanks for convincing me :)